File: test/ModularPipelines.UnitTests/Attributes/LifecycleEventIntegrationTests.cs
using Microsoft.Extensions.DependencyInjection;
using ModularPipelines.Attributes.Events;
using ModularPipelines.Configuration;
using ModularPipelines.Conditions;
using ModularPipelines.Context;
using ModularPipelines.Models;
using ModularPipelines.Modules;
using ModularPipelines.Options;
using ModularPipelines.TestHelpers;

namespace ModularPipelines.UnitTests.Attributes;

[NotInParallel(nameof(LifecycleEventIntegrationTests))]
public class LifecycleEventIntegrationTests : TestBase
{
    private static readonly List<string> EventLog = new();

    public class LogStartAttribute : Attribute, IModuleStartHandler
    {
        public Task OnModuleStartAsync(IModuleHookContext context)
        {
            EventLog.Add($"Start:{context.ModuleName}");
            return Task.CompletedTask;
        }
    }

    public class LogEndAttribute : Attribute, IModuleEndHandler
    {
        public Task OnModuleEndAsync(IModuleHookContext context, IModuleResult result)
        {
            EventLog.Add($"End:{context.ModuleName}");
            return Task.CompletedTask;
        }
    }

    public class LogFailedAttribute : Attribute, IModuleFailureHandler
    {
        public bool ContinueOnError => true;

        public Task OnModuleFailureAsync(IModuleHookContext context, Exception exception)
        {
            EventLog.Add($"Failed:{context.ModuleName}:{exception.Message}");
            return Task.CompletedTask;
        }
    }

    public class LogSkippedAttribute : Attribute, IModuleSkippedHandler
    {
        public Task OnModuleSkippedAsync(IModuleHookContext context, SkipDecision reason)
        {
            EventLog.Add($"Skipped:{context.ModuleName}:{reason.Reason}");
            return Task.CompletedTask;
        }
    }

    public class AlwaysTrueCondition : IRunCondition
    {
        public Task<bool> EvaluateAsync(IPipelineContext context) => Task.FromResult(true);
    }

    [LogStart]
    [LogEnd]
    public class SuccessfulModule : Module<string>
    {
        protected internal override async Task<string> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.Yield();
            return "Success";
        }
    }

    [LogStart]
    [LogFailed]
    public class FailingModule : Module<string>
    {
        protected internal override Task<string> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            throw new InvalidOperationException("Intentional failure");
        }
    }

    [LogStart]
    [LogSkipped]
    public class SkippingModule : Module<string>
    {
        protected override ModuleConfiguration Configure() => ModuleConfiguration.Create()
            .WithSkipWhen(_ => SkipDecision.Skip("Test skip reason"))
            .Build();

        protected internal override Task<string> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            return Task.FromResult<string>("Should not execute");
        }
    }

    [ModularPipelines.Attributes.SkipIf<AlwaysTrueCondition>]
    [LogStart]
    [LogSkipped]
    public class AttributeSkippingModule : Module<string>
    {
        protected internal override Task<string> ExecuteAsync(
            IModuleContext context,
            CancellationToken cancellationToken)
        {
            return Task.FromResult<string>("Should not execute");
        }
    }

    [Before(Test)]
    public void ClearEventLog()
    {
        EventLog.Clear();
    }

    [Test]
    public async Task SuccessfulModule_InvokesStartAndEndEvents()
    {
        var result = await TestPipelineBuilder.Create()
            .AddModule<SuccessfulModule>()
            .RunAsync();

        await Assert.That(result.Status).IsEqualTo(Enums.ModuleStatus.Succeeded);
        await Assert.That(EventLog).Contains("Start:SuccessfulModule");
        await Assert.That(EventLog).Contains("End:SuccessfulModule");
    }

    [Test]
    public async Task FailingModule_InvokesStartAndFailedEvents()
    {
        try
        {
            await TestPipelineBuilder.Create()
                .AddModule<FailingModule>()
                .ConfigurePipelineOptions((_, options) => options with
                {
                    ExecutionMode = ExecutionMode.WaitForAllModules,
                })
                .RunAsync();
        }
        catch
        {
            // Expected - the pipeline throws when a module fails
        }

        await Assert.That(EventLog).Contains("Start:FailingModule");
        await Assert.That(EventLog.Any(e => e.StartsWith("Failed:FailingModule:"))).IsTrue();
    }

    [Test]
    public async Task SkippingModule_InvokesStartAndSkippedEvents()
    {
        var result = await TestPipelineBuilder.Create()
            .AddModule<SkippingModule>()
            .RunAsync();

        await Assert.That(result.Status).IsEqualTo(Enums.ModuleStatus.Succeeded);
        await Assert.That(EventLog).Contains("Start:SkippingModule");
        await Assert.That(EventLog.Any(e => e.Contains("Skipped:SkippingModule:Test skip reason"))).IsTrue();
    }

    [Test]
    public async Task AttributeSkippingModule_InvokesStartAndSkippedEvents()
    {
        var result = await TestPipelineBuilder.Create()
            .AddModule<AttributeSkippingModule>()
            .RunAsync();

        await Assert.That(result.Status).IsEqualTo(Enums.ModuleStatus.Succeeded);
        await Assert.That(EventLog).Contains("Start:AttributeSkippingModule");
        await Assert.That(EventLog.Any(e => e.Contains("Skipped:AttributeSkippingModule:SkipIf<AlwaysTrueCondition> returned true"))).IsTrue();
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Attributes/LifecycleEventIntegrationTests.fs
namespace ModularPipelines.UnitTests.FSharp.Attributes

open System
open System.Threading
open System.Threading.Tasks
open ModularPipelines.Attributes.Events
open ModularPipelines.Configuration
open ModularPipelines.Context
open ModularPipelines.Enums
open ModularPipelines.Extensions
open ModularPipelines.Models
open ModularPipelines.Modules
open ModularPipelines.Options
open ModularPipelines.TestHelpers
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core

module LifecycleEventIntegrationTests =
    let eventLog = ResizeArray<string>()

    type LogStartAttribute() =
        inherit Attribute()
        interface IModuleStartHandler with
            member _.ContinueOnError = false
            member _.OnModuleStartAsync(context: IModuleHookContext) =
                eventLog.Add($"Start:{context.ModuleName}")
                Task.CompletedTask

    type LogEndAttribute() =
        inherit Attribute()
        interface IModuleEndHandler with
            member _.ContinueOnError = false
            member _.OnModuleEndAsync(context: IModuleHookContext, _: IModuleResult) =
                eventLog.Add($"End:{context.ModuleName}")
                Task.CompletedTask

    type LogFailedAttribute() =
        inherit Attribute()
        interface IModuleFailureHandler with
            member _.ContinueOnError = true
            member _.OnModuleFailureAsync(context: IModuleHookContext, ex: Exception) =
                eventLog.Add($"Failed:{context.ModuleName}:{ex.Message}")
                Task.CompletedTask

    type LogSkippedAttribute() =
        inherit Attribute()
        interface IModuleSkippedHandler with
            member _.ContinueOnError = false
            member _.OnModuleSkippedAsync(context: IModuleHookContext, reason: SkipDecision) =
                eventLog.Add($"Skipped:{context.ModuleName}:{reason.Reason}")
                Task.CompletedTask

    [<LogStart>]
    [<LogEnd>]
    type SuccessfulModule() =
        inherit Module<string>()
        override _.ExecuteAsync(_: IModuleContext, _: CancellationToken) =
            task {
                do! Task.Yield()
                return "Success"
            }

    [<LogStart>]
    [<LogFailed>]
    type FailingModule() =
        inherit Module<string>()
        override _.ExecuteAsync(_: IModuleContext, _: CancellationToken) : Task<string> =
            raise (InvalidOperationException("Intentional failure"))

    [<LogStart>]
    [<LogSkipped>]
    type SkippingModule() =
        inherit Module<string>()
        override _.Configure() =
            ModuleConfiguration.Create()
                .WithSkipWhen(fun () -> SkipDecision.Skip("Test skip reason"))
                .Build()
        override _.ExecuteAsync(_: IModuleContext, _: CancellationToken) =
            Task.FromResult<string>("Should not execute")

[<NotInParallel(nameof LifecycleEventIntegrationTests)>]
type LifecycleEventIntegrationTests() =
    inherit TestBase()

    [<Before(HookType.Test)>]
    member _.ClearEventLog() =
        LifecycleEventIntegrationTests.eventLog.Clear()

    [<Test>]
    member _.SuccessfulModule_InvokesStartAndEndEvents() = async {
        let! result =
            TestPipelineHostBuilder.Create()
                .AddModule<LifecycleEventIntegrationTests.SuccessfulModule>()
                .ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(result.Status = Status.Successful).IsTrue())
        do! check(Assert.That(LifecycleEventIntegrationTests.eventLog |> Seq.contains "Start:SuccessfulModule").IsTrue())
        do! check(Assert.That(LifecycleEventIntegrationTests.eventLog |> Seq.contains "End:SuccessfulModule").IsTrue())
    }

    [<Test>]
    member _.FailingModule_InvokesStartAndFailedEvents() = async {
        try
            do!
                TestPipelineHostBuilder.Create()
                    .AddModule<LifecycleEventIntegrationTests.FailingModule>()
                    .ConfigurePipelineOptions(fun _ options ->
                        options.ExecutionMode <- ExecutionMode.WaitForAllModules)
                    .ExecutePipelineAsync()
                |> Async.AwaitTask
                |> Async.Ignore
        with _ -> ()

        do! check(Assert.That(LifecycleEventIntegrationTests.eventLog |> Seq.contains "Start:FailingModule").IsTrue())
        do! check(Assert.That(LifecycleEventIntegrationTests.eventLog |> Seq.exists (fun e -> e.StartsWith("Failed:FailingModule:"))).IsTrue())
    }

    [<Test>]
    member _.SkippingModule_InvokesStartAndSkippedEvents() = async {
        let! result =
            TestPipelineHostBuilder.Create()
                .AddModule<LifecycleEventIntegrationTests.SkippingModule>()
                .ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(result.Status = Status.Successful).IsTrue())
        do! check(Assert.That(LifecycleEventIntegrationTests.eventLog |> Seq.contains "Start:SkippingModule").IsTrue())
        do! check(Assert.That(LifecycleEventIntegrationTests.eventLog |> Seq.exists (fun e -> e.Contains("Skipped:SkippingModule:Test skip reason"))).IsTrue())
    }
